### PR TITLE
[ARCTIC-561] Fix 561. When flink job fails over, flink job will commit a wrong transaction id

### LINKED ISSUE #561: [Bug]: When flink job fails over, flink job will commit a wrong transaction id
### What happened?

When flink job fail over, flink job will commit a wrong transaction id. 
The wrong transaction id is 
image
And the right transaction id is 
image
The data associated with the wrong transaction id will be lost.
The problem can only be resolved by rebuilding the table and rewriting the data.

### Affects Versions

0.3.x

### What engines are you seeing the problem on?

Flink

### How to reproduce

Insert data through flink and then make flink job failover

### Relevant log output

_No response_

### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by zstraw**
## Why are the changes needed?
FIX #561 

## Brief change log

  - *The flink 1.12 and 1.14 , 1.15versions are affected.*

## How was this patch tested?
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zhoujinsong** [comment]
@zstraw Hi, Could you please add some description about the cause and solution of this bug.

**Comment by zstraw** [comment]
> @zstraw Hi, Could you please add some description about the cause and solution of this bug.

The cause of this bug is that, transactionId is applied by jobId+checkpointId in flink file writer , and it would be the same if jobId+checkpointId is identical. When flink job failovers, the checkpointId would be initilized by value 0, which makes the applied transactionId is identical with the first transactionId in the job runs for the first time.

TransactionId can be applied by UUID+jobId+ckpId, so that transactionId can be identical once the job failovers or restarts.

**Comment by YesOrNo828** [comment]
Why could the client apply the same transaction twice if this transaction has been committed?
This question came to my mind. Could anyone have a reasonable explanation?

**Comment by zhoujinsong** [comment]
> Why could the client apply the same transaction twice if this transaction has been committed? This question came to my mind. Do I miss the context? Could anyone have a reasonable explanation?

Row sequence number is composed of transaction id and row offset, wich is used to compare bettwen different rows when optimizing and MOR processing, So the transaction id is required to be incresement.

**Review by zhoujinsong — APPROVED** [review]
LGTM.

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 1]
```diff
@@ -126,6 +157,11 @@ private Long getTransactionId() {
     return transaction;
```
Comment (YesOrNo828) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:157]: Maybe adding a log message to show how to produce the signature is better. So it is clear what combines the signature.

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java [hunk 2]
```diff
@@ -104,6 +109,32 @@ public void open() {
...
+    checkpointState =
+        context.getOperatorStateStore()
+            .getListState(
+                new ListStateDescriptor<>(
+                    subTaskId + "-task-file-writer-state",
+                    LongSerializer.INSTANCE));
+    
+    if (context.isRestored()) {
+      // get last ckp num from state when failover continuously
+      checkpointId = checkpointState.get().iterator().next();
```
Comment (YesOrNo828) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/ArcticFileWriter.java:126]: Should the value of `checkpointId` plus one?

**Review by YesOrNo828 — APPROVED** [review]

**Pull request merged**